jena-rdfconnection/src/main/java/org/apache/jena/rdflink/LibRDFLink.java

```diff
@@ -0,0 +1,109 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jena.rdflink;
+
+import org.apache.jena.graph.Graph;
+import org.apache.jena.graph.Node;
+import org.apache.jena.graph.NodeFactory;
+import org.apache.jena.http.HttpLib;
+import org.apache.jena.query.Dataset;
+import org.apache.jena.query.DatasetFactory;
+import org.apache.jena.rdf.model.Model;
+import org.apache.jena.rdf.model.ModelFactory;
+import org.apache.jena.sparql.core.DatasetGraph;
+import org.apache.jena.sparql.core.Quad;
+
+/** package-wide utilities etc */
+class LibRDFLink {
+    private static String dftName =  "default" ;
+
+    /*package*/ static boolean isDefault(Node name) {
+        return name == null || Quad.isDefaultGraph(name);
+    }
+
+    private static String queryStringForGraph(String ch, Node graphName) {
+        return
+            ch +
+                (LibRDFLink.isDefault(graphName)
+                ? "default"
+                : "graph="+encode(graphName));
+    }
+
+    private static String encode(Node node) {
+        return HttpLib.urlEncodeQueryString(node.getURI());
+    }
+
+    /*package*/ static String urlForGraph(String graphStoreProtocolService, Node graphName) {
+        // If query string
+        String ch = "?";
+        if ( graphStoreProtocolService.contains("?") )
+            // Already has a query string, append with "&"
+            ch = "&";
+        return graphStoreProtocolService + queryStringForGraph(ch, graphName);
+    }
+
+    /*package*/ static Model graph2model(Graph graph) {
+        return ModelFactory.createModelForGraph(graph);
+    }
+
+    /*package*/ static Graph model2graph(Model model) {
+        return model.getGraph();
+    }
+
+    /*package*/ static Dataset asDataset(DatasetGraph dsg) {
+        return DatasetFactory.wrap(dsg);
+    }
+
+    /*package*/ static DatasetGraph asDatasetGraph(Dataset dataset) {
+        return dataset.asDatasetGraph();
+    }
+
+    /*package*/ static Node name(String graphName) {
+        if ( graphName == null || graphName.equals("default") )
+            return Quad.defaultGraphIRI;
+        return NodeFactory.createURI(graphName);
+    }
+
+    /*package*/ static String formServiceURL(String destination, String srvEndpoint) {
+        if ( srvEndpoint == null )
+            return null;
+        if ( srvEndpoint == RDFLinkHTTPBuilder.SameAsDestination )
+            return destination;
+        if ( destination == null )
+            return srvEndpoint;
+
+        // If the srvEndpoint looks like an absolute URL, use as given.
+        if ( srvEndpoint.startsWith("http:/") || srvEndpoint.startsWith("https:/") )
```

Comment:
I assume the URL is validated elsewhere, so `http:/localhost:3030/....` won't create any issues here (i.e. user forgot and used just 1 slash). It will probably fail elsewhere first or it's harmless I think?

Reply:
Yes - it'll fail later. It protects the following code from the (bizarre) URL `http:/path` which is legal as IRI (!!) - which has no host.